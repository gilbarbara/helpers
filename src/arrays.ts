import { PlainObject, SortFunction } from './types';

/**
 * Sort an array with localeCompare
 */
export function sortByLocaleCompare(
  key?: string,
  options: Intl.CollatorOptions & { descending?: boolean } = {},
): SortFunction {
  const { descending, ...compareOptions } = options;

  if (key) {
    if (descending) {
      return <T extends PlainObject>(left: T, right: T) =>
        right[key].toLowerCase().localeCompare(left[key].toLowerCase(), undefined, compareOptions);
    }

    return <T extends PlainObject>(left: T, right: T) =>
      left[key].toLowerCase().localeCompare(right[key].toLowerCase(), undefined, compareOptions);
  }

  if (descending) {
    return <T extends string>(left: T, right: T) =>
      right.toLowerCase().localeCompare(left.toLowerCase(), undefined, compareOptions);
  }

  return <T extends string>(left: T, right: T) =>
    left.toLowerCase().localeCompare(right.toLowerCase(), undefined, compareOptions);
}

/**
 * Sort an array by primitive values
 */
export function sortByPrimitive<T extends number | boolean>(
  key?: string,
  descending = false,
): SortFunction<T> {
  const firstComparator = descending ? 1 : -1;
  const secondComparator = descending ? -1 : 1;

  if (key) {
    return <P extends PlainObject>(left: P, right: P) => {
      if (left[key] === right[key]) {
        return 0;
      }

      return left[key] < right[key] ? firstComparator : secondComparator;
    };
  }

  return <P extends T>(left: P, right: P) => {
    if (left === right) {
      return 0;
    }

    return left < right ? firstComparator : secondComparator;
  };
}
