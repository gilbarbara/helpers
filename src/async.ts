import 'isomorphic-fetch';

import { CorsOptions, RequestError, RequestOptions } from './types';

/**
 * Format a CORS response
 */
export function cors(data: any, statusCode = 200, options?: CorsOptions) {
  const { methods = ['GET'], origin = '*', headers = [] } = options || {};
  const allowMethods = [...methods, 'OPTIONS'];
  const allowHeaders = Array.from(
    new Set([
      'Accept-Version',
      'Accept',
      'Authorization',
      'Content-Length',
      'Content-MD5',
      'Content-Type',
      'Date',
      'x-amz-date',
      'x-amz-security-token',
      'X-Api-Version',
      'X-CSRF-Token',
      'X-Requested-With',
      ...headers,
    ]),
  );

  return {
    body: JSON.stringify(data),
    headers: {
      'Access-Control-Allow-Origin': origin,
      'Access-Control-Allow-Credentials': true,
      'Access-Control-Allow-Methods': allowMethods.join(','),
      'Access-Control-Allow-Headers': allowHeaders.join(','),
    },
    statusCode,
  };
}

/**
 * Make async requests
 */
export async function request(url: string, options: RequestOptions = {}): Promise<any> {
  const { body, headers, method = 'GET' }: RequestOptions = options;

  if (!url) {
    throw new Error('URL is required');
  }

  const params: RequestOptions = {
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      ...headers,
    },
    method,
  };

  if (body) {
    params.body = JSON.stringify(body);
  }

  return fetch(url, params).then(async response => {
    const text = await response.text();
    let content: any;

    try {
      content = JSON.parse(text);
    } catch {
      content = text;
    }

    if (response.status > 299) {
      const error = new Error(response.statusText) as RequestError;

      error.status = response.status;
      error.response = content;

      throw error;
    } else {
      return content;
    }
  });
}

/**
 * Block execution
 */
export function sleep(seconds = 1) {
  return new Promise(resolve => setTimeout(resolve, seconds * 1000));
}
