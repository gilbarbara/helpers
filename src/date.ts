import is from 'is-lite';

import { pluralize } from './strings';
import { TimeSinceOptions } from './types';

export const MINUTE = 60;
export const HOUR = MINUTE * 60;
export const DAY = HOUR * 24;
export const WEEK = DAY * 7;
export const MONTH = DAY * 30;
export const YEAR = 365 * DAY;

export function isIsoDate(input: string) {
  if (!/\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{3}Z/.test(input)) {
    return false;
  }

  const date = new Date(input);

  return date.toISOString() === input;
}

/**
 * Check if the input is a valid date.
 */
export function isoDate(input?: string | number) {
  if (typeof input !== 'undefined') {
    return new Date(input).toISOString();
  }

  return new Date().toISOString();
}

/**
 * Check if the input is a valid date.
 */
export function isValidDate(input: string | number | Date) {
  const date = is.date(input) ? input : new Date(input);

  return !Number.isNaN(date.getTime());
}

/**
 * Returns the unixtime (in seconds).
 */
export function now() {
  return Math.floor(Date.now() / 1000);
}

/**
 * Returns how much time has passed since the input.
 */
export function timeSince(input: Date | string | number, options?: TimeSinceOptions) {
  const {
    day = 'day',
    days,
    hour = 'hour',
    hours,
    minute = 'minute',
    minutes,
    month = 'month',
    months,
    prefix,
    second = 'second',
    seconds,
    skipWeeks = false,
    suffix = 'ago',
    week = 'week',
    weeks,
    year = 'year',
    years,
  } = options || {};
  const date = typeof input === 'number' ? input : timestamp(input);
  const diff = now() - date;
  let quantity = Math.floor(diff / YEAR);

  const output = (value: number, caption: string) => {
    let result = `${value} ${caption}`;

    if (prefix) {
      result = `${prefix} ${result}`;
    }

    if (suffix) {
      result = `${result} ${suffix}`;
    }

    return result;
  };

  if (quantity >= 1) {
    return output(quantity, pluralize(quantity, year, years));
  }

  quantity = Math.floor(diff / MONTH);

  if (quantity > 1) {
    return output(quantity, pluralize(quantity, month, months));
  }

  if (!skipWeeks) {
    quantity = Math.floor(diff / WEEK);

    if (quantity > 1) {
      return output(quantity, pluralize(quantity, week, weeks));
    }
  }

  quantity = Math.floor(diff / DAY);

  if (quantity >= 1) {
    return output(quantity, pluralize(quantity, day, days));
  }

  quantity = Math.floor(diff / HOUR);

  if (quantity >= 1) {
    return output(quantity, pluralize(quantity, hour, hours));
  }

  quantity = Math.floor(diff / MINUTE);

  if (quantity > 1) {
    return output(quantity, pluralize(quantity, minute, minutes));
  }

  return output(Math.floor(diff), pluralize(diff, second, seconds));
}

/**
 * Get the timestamp (in seconds) for a date.
 */
export function timestamp(input?: Date | string) {
  if (!input) {
    return now();
  }

  const date = typeof input === 'string' ? new Date(input) : input;

  return Math.floor(date.getTime() / 1000);
}
