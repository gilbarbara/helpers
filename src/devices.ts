/**
 * Detect if the device is in dark mode
 */
export function isDarkMode() {
  return !!window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches;
}

/**
 * Detect if the device supports touch events
 */
export function isTouchDevice() {
  return !!window.matchMedia && window.matchMedia('(pointer: coarse)').matches;
}

/**
 * Detect if the user prefers reduced motion
 */
export function prefersReducedMotion() {
  return !!window.matchMedia && window.matchMedia('prefers-reduced-motion').matches;
}
