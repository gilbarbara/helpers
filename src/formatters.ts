import { cleanupNumericString } from './strings';
import { FormatDateLocaleOptions, FormatMoneyOptions } from './types';

/**
 * Format boolean into a Yes/No string
 */
export function formatBoolean(input: boolean) {
  return input ? 'Yes' : 'No';
}

/**
 * Format string into a CPF
 */
export function formatCPF(value: string) {
  const clearValue = cleanupNumericString(value);

  if (clearValue.length >= 10) {
    return `${clearValue.slice(0, 3)}.${clearValue.slice(3, 6)}.${clearValue.slice(
      6,
      9,
    )}-${clearValue.slice(9, 11)}`;
  } else if (clearValue.length >= 7) {
    return `${clearValue.slice(0, 3)}.${clearValue.slice(3, 6)}.${clearValue.slice(6, 9)}`;
  } else if (clearValue.length >= 4) {
    return `${clearValue.slice(0, 3)}.${clearValue.slice(3, 6)}`;
  }

  return clearValue;
}

/**
 * Format date ISO string using locale
 */
export function formatDateLocale(input: string, options: FormatDateLocaleOptions = {}) {
  const { locale = 'en-GB', showTime = false } = options;

  const formatOptions: Intl.DateTimeFormatOptions = {
    year: '2-digit',
    month: '2-digit',
    day: '2-digit',
  };

  if (showTime) {
    formatOptions.hour = '2-digit';
    formatOptions.minute = '2-digit';
  }

  return new Date(input).toLocaleDateString(locale, formatOptions);
}

/**
 * Format number into money string
 */
export function formatMoney(input: number, options: FormatMoneyOptions = {}) {
  const { decimalChar = '.', showCents = false, symbol = '$', thousandsChar = ',' } = options;

  const isNegative = input < 0;
  const value = Math.abs(input);
  const [amount, cents] = value.toFixed(2).split('.');
  const padStart = amount.length > 3 ? amount.length % 3 : 0;
  const initial = amount.slice(0, padStart);
  const remain = amount.slice(padStart).replace(/(\d{3})(?=\d)/g, `$1${thousandsChar}`);

  let formatted = `${remain}`;

  if (initial) {
    formatted = `${initial}${thousandsChar}${remain}`;
  }

  if (cents !== '00' || showCents) {
    formatted += `${decimalChar}${cents}`;
  }

  return `${isNegative ? '-' : ''}${symbol}${formatted}`;
}

/**
 * Format string into a brazilian phone
 */
export function formatPhoneBR(input: string) {
  const phone = input.replace(/\D/g, '');

  if (phone.length === 8) {
    return phone.replace(/^(\d{4})(\d{4}).*/, '$1-$2');
  }

  if (phone.length === 9) {
    return phone.replace(/^(\d{5})(\d{4}).*/, '$1-$2');
  }

  if (phone.length === 10 || phone.length === 11) {
    return phone.replace(/^(\d{2})(\d{4,5})(\d{4}).*/, '($1) $2-$3');
  }

  return phone;
}

/**
 * Format string into a US phone
 */
export function formatPhoneUS(input: string) {
  const phone = input.replace(/\D/g, '');

  if (phone.length === 10) {
    return phone.replace(/^(\d{3})(\d{3})(\d{4}).*/, '($1) $2-$3');
  }

  if (phone.length === 11 && phone.startsWith('1')) {
    return phone.replace(/^1(\d{3})(\d{3})(\d{4}).*/, '+1 ($1) $2-$3');
  }

  return phone;
}

/**
 * Format string into a zip code
 */
export function formatPostalCodeBR(value: string) {
  const clearValue = cleanupNumericString(value);

  if (clearValue.length >= 6) {
    return `${clearValue.slice(0, 5)}-${clearValue.slice(5, 8)}`;
  }

  return clearValue;
}
