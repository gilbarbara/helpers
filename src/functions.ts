/**
 * Decouple methods from objects
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export function demethodize(fn: Function) {
  return (parameter: any, ...rest: any[]) => fn.apply(parameter, rest);
}

/**
 * Measure function execution time
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export async function measureExecutionTime<T = any>(callback: Function): Promise<T> {
  const start = performance.now();

  const result = await callback();

  const end = performance.now();
  const total = end - start;

  // eslint-disable-next-line no-console
  console.log(`Completed in ${Math.ceil(total)} milliseconds`);

  return result as T;
}

/**
 * A function that does nothing.
 */
export function noop() {
  return undefined;
}

/**
 * Creates a function that will only be called once.
 * Repeat calls return the value of the first invocation.
 */
export function once<T extends (...arguments_: Array<any>) => any>(fn: T): T {
  let done = false;
  let result: ReturnType<T>;

  return function Fn(...arguments_: Parameters<T>): ReturnType<T> {
    if (!done) {
      done = true;
      result = fn(...arguments_);
    }

    return result as ReturnType<T>;
  } as T;
}

/**
 * Combine multiple functions into one.
 * The output of each function is passed as the input to the next.
 */
export function pipe<T>(...fns: Array<(argument: T) => T>) {
  return (input: T) => fns.reduce((previousValue, fn) => fn(previousValue), input);
}
