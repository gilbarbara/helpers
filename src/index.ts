export * from './arrays';
export * from './async';
export * from './misc';
export * from './numbers';
export * from './objects';
export * from './strings';

export * from './types';
