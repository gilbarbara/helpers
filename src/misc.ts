import { pad } from './numbers';
import { LoggerOptions, UniqueOptions } from './types';

/**
 * Copy a string to the clipboard
 */
export async function copyToClipboard(input: string) {
  try {
    await navigator.clipboard.writeText(input);
  } catch {
    return false;
  }

  return true;
}

/**
 * Decouple methods from objects
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export function demethodize(fn: Function) {
  return (parameter: any, ...rest: any[]) => fn.apply(parameter, rest);
}

/**
 * Check if a string is a valid JSON
 */
export function isJSON(input: string) {
  try {
    JSON.parse(input);
  } catch {
    return false;
  }

  return true;
}

/**
 * Throw an error if the parameter isn't provided
 */
export function isRequired(input = 'parameter', Constructable = TypeError) {
  throw new Constructable(`"${input}" is required`);
}

/**
 * Log grouped messages to the console
 */
export function logger(type: string, title: string, data: any, options: LoggerOptions = {}) {
  /* eslint-disable no-console */
  const { collapsed = true, hideTimestamp = false, skip = false, typeColor = 'gray' } = options;
  const groupMethod = collapsed ? console.groupCollapsed : console.group;
  const date = new Date();
  const parts = [`%c ${type}`];
  const styles = [`color: ${typeColor}; font-weight: lighter;`, 'color: inherit;'];

  if (!hideTimestamp) {
    styles.push('color: gray; font-weight: lighter;');
  }

  const time = `${pad(date.getHours(), 2)}:${pad(date.getMinutes(), 2)}:${pad(
    date.getSeconds(),
    2,
  )}`;

  parts.push(`%c${title}`);

  if (!hideTimestamp) {
    parts.push(`%c@ ${time}`);
  }

  /* istanbul ignore else */
  if (!skip) {
    groupMethod(parts.join(' '), ...styles);
    console.log(data);
    console.groupEnd();
  }
  /* eslint-enable */
}

/**
 * Measure function execution time
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export async function measureExecutionTime<T = any>(callback: Function): Promise<T> {
  const start = performance.now();

  const result = await callback();

  const end = performance.now();
  const total = end - start;

  // eslint-disable-next-line no-console
  console.log(`Completed in ${Math.ceil(total)} milliseconds`);

  return result as T;
}

/**
 * A function that does nothing.
 */
export function noop() {
  return undefined;
}

/**
 * Returns the value or null
 */
export function nullify<T>(value: T) {
  return value ?? null;
}

export function popupCenter(url: string, title: string, width: number, height: number) {
  const { screen } = window;
  let { screenLeft } = window;
  let { screenTop } = window;
  let screenWidth = window.innerWidth;
  let screenHeight = window.innerHeight;

  /* istanbul ignore next */
  if (typeof screenLeft === 'undefined') {
    // @ts-ignore
    screenLeft = screen.left;
  }

  /* istanbul ignore next */
  if (typeof screenTop === 'undefined') {
    // @ts-ignore
    screenTop = screen.top;
  }

  /* istanbul ignore next */
  if (!screenWidth) {
    screenWidth = document.documentElement.clientWidth
      ? document.documentElement.clientWidth
      : screen.width;
  }

  /* istanbul ignore next */
  if (!screenHeight) {
    screenHeight = document.documentElement.clientHeight
      ? document.documentElement.clientHeight
      : screen.height;
  }

  const left = screenWidth / 2 - width / 2 + screenLeft;
  const top = screenHeight / 2 - height / 2 + screenTop;

  const popup = window.open(
    url,
    title,
    `menubar=no,location=no,resizable=no,scrollbars=yees,status=no,width=${width},height=${height},top=${top}, left=${left}`,
  );

  /* istanbul ignore else */
  if (popup) {
    popup.focus();
  }

  return popup;
}

/**
 * Return a unique string
 */
export function unique(length = 8, options: UniqueOptions = {}): string {
  const {
    includeLowercase = true,
    includeNumbers = true,
    includeSymbols = false,
    includeUppercase = true,
  } = options;
  const lowercase = 'abcdefghijklmnopqrstuvwxyz';
  const uppercase = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
  const numbers = '0123456789';
  const symbols = '!?@#$%^&*+_-=:.~';
  // const similarCharacters = /[ilLI|`oO0]/g;

  let characters = '';

  if (includeLowercase) {
    characters += lowercase;
  }

  if (includeUppercase) {
    characters += uppercase;
  }

  if (includeNumbers) {
    characters += numbers;
  }

  if (includeSymbols) {
    characters += symbols;
  }

  let result = '';

  for (let index = length; index > 0; --index) {
    result += characters[Math.round(Math.random() * (characters.length - 1))];
  }

  return result;
}

/**
 * Returns an UUID v4 string.
 */
export function uuid() {
  /* eslint-disable no-bitwise */
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
    const r = (Math.random() * 16) | 0;
    const v = c === 'x' ? r : (r & 0x3) | 0x8;

    return v.toString(16);
  });
  /* eslint-enable */
}
