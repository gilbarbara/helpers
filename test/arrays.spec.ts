import { sortByLocaleCompare, sortByPrimitive } from '../src';

describe('sortByLocaleCompare', () => {
  it.each([
    ['portuguese', ['Mãe', 'limão', 'cachê', 'tião', 'amô', 'côncavo'], undefined],
    ['french', ['réservé', 'Premier', 'Cliché', 'communiqué', 'café', 'Adieu'], undefined],
    [
      'english',
      ['port', 'Mars', 'Car', 'cart', 'Payment', 'asylum', 'Asian'],
      { descending: true },
    ],
  ])('should sort an array in %s', (_, input, options) => {
    expect(input.sort(sortByLocaleCompare(undefined, options))).toMatchSnapshot();
  });

  it.each([
    ['ascending', [{ key: 'green' }, { key: 'amber' }, { key: 'red' }, { key: 'cyan' }], undefined],
    [
      'descending',
      [{ key: 'green' }, { key: 'amber' }, { key: 'red' }, { key: 'cyan' }],
      { descending: true },
    ],
  ])('should sort an object %s', (_, input, options) => {
    expect(input.sort(sortByLocaleCompare('key', options))).toMatchSnapshot();
  });
});

describe('sortByPrimitive', () => {
  it.each([
    ['cycle-number-asc', 'cycle', [{ cycle: 1 }, { cycle: 4 }, { cycle: 3 }, { cycle: 3 }], false],
    ['cycle-number-desc', 'cycle', [{ cycle: 1 }, { cycle: 4 }, { cycle: 3 }, { cycle: 3 }], true],
    ['-number-asc', undefined, [2, 4, 1, 3, 2], false],
    ['-number-desc', undefined, [2, 4, 1, 3, 2], true],
    [
      'status-boolean-asc',
      'status',
      [
        { cycle: 3, status: true },
        { cycle: 1, status: false },
        { cycle: 3, status: true },
        { cycle: 4, status: false },
      ],
      false,
    ],
    [
      'status-boolean-desc',
      'status',
      [
        { cycle: 3, status: true },
        { cycle: 1, status: false },
        { cycle: 3, status: true },
        { cycle: 4, status: false },
      ],
      true,
    ],
  ])('should sort an array by %s', (_, key, input, descending) => {
    expect(input.sort(sortByPrimitive(key, descending))).toMatchSnapshot();
  });
});
