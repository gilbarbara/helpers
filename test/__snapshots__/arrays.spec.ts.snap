// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sortByLocaleCompare should sort an array in english 1`] = `
Array [
  "port",
  "Payment",
  "Mars",
  "cart",
  "Car",
  "asylum",
  "Asian",
]
`;

exports[`sortByLocaleCompare should sort an array in french 1`] = `
Array [
  "Adieu",
  "café",
  "Cliché",
  "communiqué",
  "Premier",
  "réservé",
]
`;

exports[`sortByLocaleCompare should sort an array in portuguese 1`] = `
Array [
  "amô",
  "cachê",
  "côncavo",
  "limão",
  "Mãe",
  "tião",
]
`;

exports[`sortByLocaleCompare should sort an object ascending 1`] = `
Array [
  Object {
    "key": "amber",
  },
  Object {
    "key": "cyan",
  },
  Object {
    "key": "green",
  },
  Object {
    "key": "red",
  },
]
`;

exports[`sortByLocaleCompare should sort an object descending 1`] = `
Array [
  Object {
    "key": "red",
  },
  Object {
    "key": "green",
  },
  Object {
    "key": "cyan",
  },
  Object {
    "key": "amber",
  },
]
`;

exports[`sortByPrimitive should sort an array by -number-asc 1`] = `
Array [
  1,
  2,
  2,
  3,
  4,
]
`;

exports[`sortByPrimitive should sort an array by -number-desc 1`] = `
Array [
  4,
  3,
  2,
  2,
  1,
]
`;

exports[`sortByPrimitive should sort an array by cycle-number-asc 1`] = `
Array [
  Object {
    "cycle": 1,
  },
  Object {
    "cycle": 3,
  },
  Object {
    "cycle": 3,
  },
  Object {
    "cycle": 4,
  },
]
`;

exports[`sortByPrimitive should sort an array by cycle-number-desc 1`] = `
Array [
  Object {
    "cycle": 4,
  },
  Object {
    "cycle": 3,
  },
  Object {
    "cycle": 3,
  },
  Object {
    "cycle": 1,
  },
]
`;

exports[`sortByPrimitive should sort an array by status-boolean-asc 1`] = `
Array [
  Object {
    "cycle": 1,
    "status": false,
  },
  Object {
    "cycle": 4,
    "status": false,
  },
  Object {
    "cycle": 3,
    "status": true,
  },
  Object {
    "cycle": 3,
    "status": true,
  },
]
`;

exports[`sortByPrimitive should sort an array by status-boolean-desc 1`] = `
Array [
  Object {
    "cycle": 3,
    "status": true,
  },
  Object {
    "cycle": 3,
    "status": true,
  },
  Object {
    "cycle": 1,
    "status": false,
  },
  Object {
    "cycle": 4,
    "status": false,
  },
]
`;
